supervisor/process.go

```diff
@@ -0,0 +1,231 @@
+// use to retrieve the final status
+func (p *Process) Start() {
+	p.access.Lock()
+	defer p.access.Unlock()
+
+	p.status.StartTime = time.Now()
+	if err := p.cmd.Start(); err != nil {
+		// process failed to start.
+		// exit immediately
+		p.status.StopTime = time.Now()
+		p.status.Error = err
+		p.status.State = Fatal
+		close(p.done)
+
+		return
+	}
+
+	p.status.PID = p.cmd.Process.Pid
+	p.status.State = Running
+	go p.wait()
+}
+
+// wait waits for a process to finish running and sets it's final status
+func (p *Process) wait() {
+	// send the final status and close done channel to signal to
+	// goroutines the process has finished running
+	defer func() {
+		close(p.done)
+	}()
+
+	err := p.cmd.Wait()
+	now := time.Now()
+	exitCode := 0
+
+	p.access.Lock()
+	defer p.access.Unlock()
+
+	// get the exit code of the process
+	if err != nil {
+		if exitErr, ok := err.(*exec.ExitError); ok {
+			err = nil
+			exitCode = exitErr.ExitCode()
+
+			if status, ok := exitErr.Sys().(syscall.WaitStatus); ok {
+				if status.Signaled() {
+					err = errors.New(exitErr.Error())
+				}
+			}
+		}
+	}
+
+	// set final status of the process
+	p.status.Error = err
+	p.status.ExitCode = exitCode
+	p.status.StopTime = now
+	p.status.State = Finished
+}
+
+// GetStatus returns the current status of the process
+func (p *Process) GetStatus() Status {
+	p.access.RLock()
+	defer p.access.RUnlock()
+
+	return p.status
+}
+
+// Stop stops a process by sending it's group a SIGTERM signal
+func (p *Process) Stop(byUser bool) error {
+	p.access.Lock()
+	defer p.access.Unlock()
+
+	if p.status.State != Running {
+		return ErrFailedStopProc
+	}
+
+	p.status.StopByUser = byUser
+
+	return p.cmd.Process.Signal(syscall.SIGTERM)
+}
+
+// Signal sends an os signal to the process
+func (p *Process) Signal(sig syscall.Signal) error {
+	p.access.Lock()
+	defer p.access.Unlock()
+
+	if p.status.State != Running {
+		return ErrFailedSignalProc
+	}
+
+	return syscall.Kill(-p.cmd.Process.Pid, sig)
+}
+
+// StdOut returns the standard output streaming channel
+func (p *Process) StdOut() <-chan byte {
+	return p.streamStdOut
+}
+
+// StdErr returns the standard error streaming channel
+func (p *Process) StdErr() <-chan byte {
+	return p.streamStdErr
+}
+
+// Done returnes a channel that is closed when the process finishes running
+func (p *Process) Done() chan struct{} {
+	return p.done
+}
+
+// StreamOutput streams bytes of output through the provided channel
+type StreamOutput struct {
+	streamChan chan byte
+}
+
+// NewStreamOutput returns an instance of a StreamOuput using the provided channel
+func NewStreamOutput(streamChan chan byte) *StreamOutput {
+	s := &StreamOutput{
+		streamChan: streamChan,
+	}
+
+	return s
+}
+
+// TODO: stream bytes instead of lines
+// Write implements the io.Writer interface for StreamOutput
+func (s *StreamOutput) Write(buf []byte) (int, error) {
+	n := len(buf)
+
+	for _, b := range buf {
+		s.streamChan <- b
+	}
```

Comment:
making these channels be `chan []byte` would be much more efficient than sending 1 byte at a time

Reply:
Thanks. This has been fixed